File: modules/project/src/main/java/com/falsepattern/zigbrains/project/util/CLIUtil.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.util;

import com.falsepattern.zigbrains.project.execution.ZigCapturingProcessHandler;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.execution.process.CapturingProcessHandler;
import com.intellij.execution.process.ProcessOutput;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import lombok.val;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.StringTokenizer;

public class CLIUtil {
    public static Optional<ProcessOutput> execute(GeneralCommandLine cli, int timeoutMillis) {
        val handler = ZigCapturingProcessHandler.startProcess(cli);
        return handler.map(h -> runProcessWithGlobalProgress(h, timeoutMillis));
    }

    public static ProcessOutput runProcessWithGlobalProgress(CapturingProcessHandler handler, @Nullable Integer timeoutMillis) {
        return runProcess(handler, ProgressManager.getGlobalProgressIndicator(), timeoutMillis);
    }

    public static ProcessOutput runProcess(CapturingProcessHandler handler, @Nullable ProgressIndicator indicator, @Nullable Integer timeoutMillis) {
        if (indicator != null && timeoutMillis != null) {
            return handler.runProcessWithProgressIndicator(indicator, timeoutMillis);
        } else if (indicator != null) {
            return handler.runProcessWithProgressIndicator(indicator);
        } else if (timeoutMillis != null) {
            return handler.runProcess(timeoutMillis);
        } else {
            return handler.runProcess();
        }
    }

    //From Apache Ant
    /**
     * Crack a command line.
     * @param toProcess the command line to process.
     * @return the command line broken into strings.
     * An empty or null toProcess parameter results in a zero sized array.
     */
    public static String[] translateCommandline(String toProcess) throws ConfigurationException {
        if (toProcess == null || toProcess.isEmpty()) {
            //no command? no string
            return new String[0];
        }
        // parse with a simple finite state machine

        final int normal = 0;
        final int inQuote = 1;
        final int inDoubleQuote = 2;
        int state = normal;
        final StringTokenizer tok = new StringTokenizer(toProcess, "\"' ", true);
        final ArrayList<String> result = new ArrayList<>();
        final StringBuilder current = new StringBuilder();
        boolean lastTokenHasBeenQuoted = false;

        while (tok.hasMoreTokens()) {
            String nextTok = tok.nextToken();
            switch (state) {
                case inQuote:
                    if ("'".equals(nextTok)) {
                        lastTokenHasBeenQuoted = true;
                        state = normal;
                    } else {
                        current.append(nextTok);
                    }
                    break;
                case inDoubleQuote:
                    if ("\"".equals(nextTok)) {
                        lastTokenHasBeenQuoted = true;
                        state = normal;
                    } else {
                        current.append(nextTok);
                    }
                    break;
                default:
                    if ("'".equals(nextTok)) {
                        state = inQuote;
                    } else if ("\"".equals(nextTok)) {
                        state = inDoubleQuote;
                    } else if (" ".equals(nextTok)) {
                        if (lastTokenHasBeenQuoted || current.length() > 0) {
                            result.add(current.toString());
                            current.setLength(0);
                        }
                    } else {
                        current.append(nextTok);
                    }
                    lastTokenHasBeenQuoted = false;
                    break;
            }
        }
        if (lastTokenHasBeenQuoted || current.length() > 0) {
            result.add(current.toString());
        }
        if (state == inQuote || state == inDoubleQuote) {
            throw new ConfigurationException("unbalanced quotes in " + toProcess);
        }
        return result.toArray(new String[0]);
    }

    public static List<String> colored(boolean colored, boolean debug) {
        // TODO remove this check once JetBrains implements colored terminal in the debugger
        // https://youtrack.jetbrains.com/issue/CPP-11622/ANSI-color-codes-not-honored-in-Debug-Run-Configuration-output-window
        if (debug) {
            return Collections.emptyList();
        } else {
            return List.of("--color", colored ? "on" : "off");
        }
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/shared/cli/CLIUtil.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.shared.cli

import com.falsepattern.zigbrains.ZigBrainsBundle
import com.falsepattern.zigbrains.project.run.ZigProcessHandler
import com.falsepattern.zigbrains.shared.ipc.IPCUtil
import com.falsepattern.zigbrains.shared.ipc.ipc
import com.intellij.execution.ExecutionException
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.ProcessOutput
import com.intellij.execution.process.ProcessTerminatedListener
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.project.Project
import com.intellij.util.io.awaitExit
import com.intellij.util.system.OS
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import java.nio.file.Path
import java.util.*
import kotlin.io.path.exists
import kotlin.io.path.isDirectory
import kotlin.io.path.pathString


//From Apache Ant
/**
 * Crack a command line.
 * @param toProcess the command line to process.
 * @return the command line broken into strings.
 * An empty or null toProcess parameter results in a zero sized array.
 */
@Throws(ConfigurationException::class)
fun translateCommandline(toProcess: String): List<String> {
    if (toProcess.isEmpty()) {
        //no command? no string
        return listOf()
    }

    // parse with a simple finite state machine
    val normal = 0
    val inQuote = 1
    val inDoubleQuote = 2
    var state = normal
    val tok = StringTokenizer(toProcess, "\"' ", true)
    val result = ArrayList<String>()
    val current = StringBuilder()
    var lastTokenHasBeenQuoted = false

    while (tok.hasMoreTokens()) {
        val nextTok: String = tok.nextToken()
        when (state) {
            inQuote -> if ("'" == nextTok) {
                lastTokenHasBeenQuoted = true
                state = normal
            } else {
                current.append(nextTok)
            }

            inDoubleQuote -> if ("\"" == nextTok) {
                lastTokenHasBeenQuoted = true
                state = normal
            } else {
                current.append(nextTok)
            }

            else -> {
                if ("'" == nextTok) {
                    state = inQuote
                } else if ("\"" == nextTok) {
                    state = inDoubleQuote
                } else if (" " == nextTok) {
                    if (lastTokenHasBeenQuoted || current.isNotEmpty()) {
                        result.add(current.toString())
                        current.setLength(0)
                    }
                } else {
                    current.append(nextTok)
                }
                lastTokenHasBeenQuoted = false
            }
        }
    }
    if (lastTokenHasBeenQuoted || current.isNotEmpty()) {
        result.add(current.toString())
    }
    if (state == inQuote || state == inDoubleQuote) {
        throw ConfigurationException(ZigBrainsBundle.message("exception.translate-command-line.unbalanced-quotes", toProcess))
    }
    return result
}

fun createCommandLineSafe(
    workingDirectory: Path?,
    exe: Path,
    vararg parameters: String,
): Result<GeneralCommandLine> {
    if (!exe.exists())
        return Result.failure(IllegalArgumentException("file does not exist: ${exe.pathString}"))
    if (exe.isDirectory())
        return Result.failure(IllegalArgumentException("file is a directory: ${exe.pathString}"))
    val cli = GeneralCommandLine()
        .withExePath(exe.toString())
        .withWorkDirectory(workingDirectory?.toFile())
        .withParameters(*parameters)
        .withCharset(Charsets.UTF_8)
    return Result.success(cli)
}

@Throws(ExecutionException::class)
suspend fun GeneralCommandLine.startIPCAwareProcess(project: Project?, emulateTerminal: Boolean = false): ZigProcessHandler.IPCAware {
    val original = this.commandLineString
    val ipc = if (project != null && !emulateTerminal) IPCUtil.wrapWithIPC(this) else null
    val cli = ipc?.cli ?: this
    if (emulateTerminal && OS.CURRENT != OS.Windows && !cli.environment.contains("TERM")) {
        cli.withEnvironment("TERM", "xterm-256color")
    }
    val handler = ZigProcessHandler.IPCAware(original, cli)
    ProcessTerminatedListener.attach(handler)

    if (ipc != null) {
        project!!.ipc?.launchWatcher(ipc, handler.process)
    }
    return handler
}


suspend fun GeneralCommandLine.call(timeoutMillis: Long = Long.MAX_VALUE, ipcProject: Project? = null): Result<ProcessOutput> {
    val (process, exitCode) = withContext(Dispatchers.IO) {
        val handler = startIPCAwareProcess(ipcProject)
        val process = handler.process
        val exit = withTimeoutOrNull(timeoutMillis) {
            process.awaitExit()
        }
        process to exit
    }
    return runInterruptible {
        Result.success(ProcessOutput(
            process.inputStream.bufferedReader().use { it.readText() },
            process.errorStream.bufferedReader().use { it.readText() },
            exitCode ?: -1,
            exitCode == null,
            false
        ))
    }
}